[3.x] Fix external images getting embedded on import

pablitar opened the pull request:
Fixes #94475

It keeps track of the external images and loads them accordingly as textures, instead of replacing them with generic image textures.

I tried to also take into account the "sampler" flags from the GLTF.

One thing I'm worried though is that maybe the import flags of the external file should override the configuration of the GLTF?

Before merging this it'd be great to think about that. I'm actually thinking that I should probably commit a change to ignore sampler configuration for external images, since those can be configured in the import settings.

AThousandShips commented on modules/gltf/gltf_document.cpp:
```diff
@@ -3234,17 +3235,24 @@ Error GLTFDocument::_parse_textures(Ref<GLTFState> p_state) {
+		if (p_state->external_images_paths.has(t->get_src_image())) {
+			tex = ResourceLoader::load(p_state->external_images_paths[t->get_src_image()]);
+		} else {
+			Ref<ImageTexture> imgTex;
+			imgTex.instance();
+			imgTex->create_from_image(p_state->images[t->get_src_image()]);
+			tex = imgTex;
```
```suggestion
			Ref<ImageTexture> img_tex;
			img_tex.instance();
			img_tex->create_from_image(p_state->images[t->get_src_image()]);
			tex = img_tex;
```
While we're here, let's fix the style here

aaronfranke reviewed (commented):
Reviewed. It makes sense to me. However, this is different from the code in master. So two comments:
* What does master do to solve the same problem? Which is better?
* The commits need to be squashed.

LunaticInAHat commented:
If I recall correctly (I believe I was the one who introduced the current behavior in both 3.6 and master), master just didn't have to jump through the same hoops, when trying to respect sampler settings. In 3.x, filter & repeat are flags set on the texture, so it was necessary to load & cache the textures within the GTLF document state, so that we could set the filter flags on them. In master, those flags are on the material, not the texture, so there's no equivalent need to have a local cache of the textures.

The changes here look like they have a plausible idea, however I am concerned that the resource loader isn't going to give us the behavior we want -- I am not familiar with its inner workings, but I wouldn't have thought that `load()` would return the handle of a new texture each time it was called. I would have imagined that it would cache the resources it loads, and that subsequent requests for an already-loaded resources would just be given a handle to the single copy in the cache. If that is the case, then I think this proposal will break texture sampling if you have a single texture being referenced from multiple places, with different filter / repeat options set.

pablitar commented:
> Reviewed. It makes sense to me. However, this is different from the code in master. So two comments:
> 
> * What does master do to solve the same problem? Which is better?
> * The commits need to be squashed.

Hey! I looked into what master does, and it seems to be quite different to what's on 3.x

Master uses an array of textures instead of an array of images. It's similar to what was before @LunaticInAHat change. But I don't understand how master takes the GLTF sampler configuration into account. I'd have to read further :(.

I don't really know what's better. I just tried to do the minimum change that solved the problem. On our side we are on a tight deadline so we went for that route (we are actually working on our forked version of Godot).

pablitar commented:
> k textur



> If I recall correctly (I believe I was the one who introduced the current behavior in both 3.6 and master), master just didn't have to jump through the same hoops, when trying to respect sampler settings. In 3.x, filter & repeat are flags set on the texture, so it was necessary to load & cache the textures within the GTLF document state, so that we could set the filter flags on them. In master, those flags are on the material, not the texture, so there's no equivalent need to have a local cache of the textures.
> 
> The changes here look like they have a plausible idea, however I am concerned that the resource loader isn't going to give us the behavior we want -- I am not familiar with its inner workings, but I wouldn't have thought that `load()` would return the handle of a new texture each time it was called. I would have imagined that it would cache the resources it loads, and that subsequent requests for an already-loaded resources would just be given a handle to the single copy in the cache. If that is the case, then I think this proposal will break texture sampling if you have a single texture being referenced from multiple places, with different filter / repeat options set.

I get it now. So master has a different model for texture sampling. Hmm, this is quite a conundrum hahaha.

Could there be a way to have a texture that is derived from another texture, but that has different sampling configurations?

LunaticInAHat commented:
> Could there be a way to have a texture that is derived from another texture, but that has different sampling configurations? I'm thinking of something like a "texture proxy".

Not that I am aware of. You can make scene-local versions of resources that you have loaded via ResourceLoader, which you would then be able to set the needed flags on without conflicting with other references to the same textures, however I think that doing so would probably have the same effect as is what is already being done -- the resource would become scene-local and no longer reference the filesystem.

Unless someone more knowledgeable of Godot's resource management knows better, it looks to me like we may be a bit backed into a corner by the architectures of the 3.x resource management system, and its renderer (with how texture filtering is managed).

If that is the case, then the best thing may be to just back my changes out of 3.x, and document it as a known limitation of Godot 3.x that GLTF imports will not respect sampler settings. That would suck, but seems much saner than breaking the link between an imported reference to an external texture, and that actual external texture. I hadn't really foreseen that consequence, when I made my changes, but it does look fairly obvious now in hindsight.

Probably something above 90% of assets are just going to be linear-mipmap-linear with repeat, anyways, so probably not too many people will get bitten by rolling back the feature. And those people should find that things will still work as they want, in 4.x.

pablitar commented:
> If that is the case, then the best thing may be to just back my changes out of 3.x, and document it as a known limitation of Godot 3.x that GLTF imports will not respect sampler settings. That would suck, but seems much saner than breaking the link between an imported reference to an external texture, and that actual external texture. I hadn't really foreseen that consequence, when I made my changes, but it does look fairly obvious now in hindsight.
> 
> Probably something above 90% of assets are just going to be linear-mipmap-linear with repeat, anyways, so probably not too many people will get bitten by rolling back the feature. And those people should find that things will still work as they want, in 4.x.

That's fine by me! If that's the decision, then I'll proceed to close this PR, and leave the revert for someone else. Please let me know!

lawnjelly commented:
> If that is the case, then the best thing may be to just back my changes out of 3.x, and document it as a known limitation of Godot 3.x that GLTF imports will not respect sampler settings.

It looks like this may need @LunaticInAHat or other familiar to make a manual revert, as there have been considerable changes since the original merge, and a git revert creates conflicts. I'm not familiar enough with any of this area to be making changes.

An alternative might be to keep most of the changes and see if patch up with this PR or a modification to this PR could get us to a working state.

pablitar commented:
> An alternative might be to keep most of the changes and see if patch up with this PR or a modification to this PR could get us to a working state. I really have to rely on you guys to decide how best to approach this.

One alternative could be to have the sampler settings be supported only for GLTF with embedded images (glb, glTF Embedded), and with GLTF + bin + textures use the import settings instead. It's a little inconsistent, but maybe it can be documented and would provide at least partial support to the sampler settings.

lawnjelly commented:
> One alternative could be to have the sampler settings be supported only for GLTF with embedded images (glb, glTF Embedded), and with GLTF + bin + textures use the import settings instead. It's a little inconsistent, but maybe it can be documented and would provide at least partial support to the sampler settings.

That sounds good to me, is it possible to modify the PR to do this?

pablitar commented:
> That sounds good to me, is it possible to modify the PR to do this? (Or work out between you guys whether this is possible?)

Done!

Should I squash the commits so it's ready to be merged in case it is approved?

AThousandShips commented:
> Should I squash the commits so it's ready to be merged in case it is approved?

Please do, isn't absolutely required until it's approved but helps to keep things clean early

pablitar committed:
GLTF: Fixed external images getting embedded on import

Added a map to keep track of external images during import, and used
that map to instance the textures using the resource loader instead of
creating a new texture from scratch

pablitar commented:
> > Should I squash the commits so it's ready to be merged in case it is approved?
> 
> Please do, isn't absolutely required until it's approved but helps to keep things clean early

Squashed :D

LunaticInAHat commented:
Everything looks like it makes sense, to me. Probably the best behavior we can reasonably accomplish, within the constraints of the 3.x architecture.

The only thing that might be outstanding would be adding a note somewhere in the manual, documenting this difference in behavior between embedded vs external ([maybe here](https://docs.godotengine.org/en/latest/tutorials/assets_pipeline/importing_3d_scenes/available_formats.html#exporting-gltf-2-0-files-from-blender-recommended)) -- it's not clear to me whether that should be done as a companion piece to this PR, or as a separate act after this PR gets merged.

lawnjelly commented:
> The only thing that might be outstanding would be adding a note somewhere in the manual, documenting this difference in behavior between embedded vs external ([maybe here](https://docs.godotengine.org/en/latest/tutorials/assets_pipeline/importing_3d_scenes/available_formats.html#exporting-gltf-2-0-files-from-blender-recommended)) -- it's not clear to me whether that should be done as a companion piece to this PR, or as a separate act after this PR gets merged.

Docs changes are separate PRs to the docs repository (classref is internal to the main engine, but even there, in some situations it can be better to have a separate PR).

@LunaticInAHat if you are able to test and give approval (if it looks okay) to the PR that would be super useful, I will try and run and test it tomorrow. If we all agree it is fine and it is approved, I can merge.

@pablitar In terms of the commit message, it might be worth simplifying as e.g. the names of the multiple commits which were squashed won't be important, just a title and a sentence or two to describe what it is doing is great.

pablitar commented:
> @pablitar In terms of the commit message, it might be worth simplifying as e.g. the names of the multiple commits which were squashed won't be important, just a title and a sentence or two to describe what it is doing is great.

Done! Hope it's OK!

LunaticInAHat commented:
I have gone ahead and tested this, using the MRP from the linked issue.

It resolved the problem with breaking the linkage between the imported material and the source textures, when imported from the GLTF. The path was recorded properly within the material, and changing the underlying image on disk updated the material.

I also imported the GLTF into Blender and re-exported it as GLB (renaming the material, so that it would get imported under a new name). As expected, importing the GLB yielded a material whose texture was embedded within it, and changes to the image on disk had no effect.

So, looks like it had the desired effect, to me.

lawnjelly reviewed (approved):
I'm not an expert in this area but @LunaticInAHat seems to be giving an approval (without a review approval) and it looks fine to me.

The major danger here (as I see it) is duplicate textures being loaded in the case of embedded GLTFs but am trusting the contributors that this should not happen.

I'll try and double check this with Akien before merging.

LunaticInAHat commented:
Sorry, hadn't formally reviewed a PR on Github before; didn't know where the Approve button was, and wasn't sure that I could even press it. I'm just a guy who made a driveby contribution to GLTF import in the past, not a Godot "import" maintainer.

As for your concern: Duplicate textures will absolutely happen, in the case of embedded imports, if the materials are not named the same across the imported assets (or if the import settings are such that materials are imported internal to the meshes, rather than written to separate files).

That duplication of textures is required, in order to be able to respect potentially-conflicting sampler settings (and is also the current behavior in 3.6). If that is undesired, then we would have to completely give up on respecting the sampler settings specified in the GLTF, for 3.x. That was the behavior of the branch prior to yesterday.

There is a judgment call that needs to be made here, about which behavior is worse:
 - Failing to respect the settings specified in the imported file, or
 - Potentially creating duplicate textures if different assets imported from GLB reference the same source image

I don't really feel sufficiently involved in the Godot project to be comfortable making that decision for the project. My gut says that creating duplicate textures is probably the worse thing for peoples' games, but I would be looking for someone more invested into the engine (and familiar with its performance characteristics) to make that call.

lawnjelly commented:
> Sorry, hadn't formally reviewed a PR on Github before; didn't know where the Approve button was, and wasn't sure that I could even press it. I'm just a guy who made a driveby contribution to GLTF import in the past, not a Godot "import" maintainer.

Approvals by non-maintainers are often super useful, especially when there is no active maintainer in an area.

> As for your concern: Duplicate textures will absolutely happen, in the case of embedded imports, if the materials are not named the same across the imported assets (or if the import settings are such that materials are imported internal to the meshes, rather than written to separate files).

This is probably expected at this stage, although could possibly be later improved (if a duplicate is detected with similar flags for example). But I'm hoping it should be possible again with this PR to intentionally import several different GLTFs and share textures (provided these are specified externally and share the same name).

LunaticInAHat commented:
> This is probably expected at this stage, although could possibly be later improved (if a duplicate is detected with similar flags for example). But I'm hoping it should be possible again with this PR to intentionally import several different GLTFs and share textures (provided these are specified externally and share the same name).

Yes. When I tested with the MRP, importing the GLTF (which referenced its textures via filename) resulted in a material that still referenced the textures on-disk, and reacted to updates to those textures on-disk. So, having multiple GLTF that all reference the same textures may result in multiple materials being created (if the material names are not the same), but those materials should still reference the on-disk texture, and share it at the ResourceLoader level.

It was only when importing a GLB (with images embedded) that I saw duplicate textures being created (as expected)

fire commented:
The logic of what happens makes sense as a design and any tests I would do would be to to manually verify that the design matches the design in practice. So try to break the design. I haven’t used godot 3 in a while so it is hard for me to get back into the testing it.

aaronfranke reviewed (approved):
I'm actually not sure what the difference is between now and when the PR first opened, but regardless looks good to me. It provides a 3.x alternative to a similar system in 4.x. Note: I have not tested running this, I just looked over the code.

akien-mga reviewed (approved):
Looks good to me.

akien-mga commented:
Thanks!
